{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[!img](http://4.bp.blogspot.com/-Ocn91gy9E5M/UBc1emNFh9I/AAAAAAAAAVY/E6vmshsdeZ4/s1600/Project+log+(BU_HD)+(1).png)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# On strand information again\n",
    "`cufflinks` supports the following library types.\n",
    "ff-firststrand\n",
    "        ff-secondstrand\n",
    "        ff-unstranded\n",
    "        fr-firststrand\n",
    "        fr-secondstrand\n",
    "        fr-unstranded (default)\n",
    "        transfrags\n",
    "\n",
    "\n",
    "fr-unstranded: Standard Illumina Reads from the left-most end of the fragment (in transcript coordinates) map to the transcript strand, and the right-most end maps to the opposite strand.\n",
    "\n",
    "fr-firststrand:dUTP, NSR, NNSR Same as above except we enforce the rule that the right-most end of the fragment (in transcript coordinates) is the first sequenced (or only sequenced for single-end reads). Equivalently, it is assumed that only the strand generated during first strand synthesis is sequenced.\n",
    "\n",
    "fr-secondstrand: Ligation, Standard SOLiD Same as above except we enforce the rule that the left-most end of the fragment (in transcript coordinates) is the first sequenced (or only sequenced for single-end reads). Equivalently, it is assumed that only the strand generated during second strand synthesis is sequenced.\n",
    "\n",
    "\n",
    "This is super counter intuitiuve. But if you have single end data and the data is forward stranded\n",
    "(where I assume that this defnition of foorward is standard: the reads map in the same \n",
    "direction as the transcripts are defined.), then you want to use `-fr-secondstrand` and \n",
    "not `-fr-firststrand` as you would otherwise assume\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEO</th>\n",
       "      <th>SRA</th>\n",
       "      <th>RIBO sample</th>\n",
       "      <th>RNA sample</th>\n",
       "      <th>read lengths</th>\n",
       "      <th>offsets</th>\n",
       "      <th>RIBO adapter</th>\n",
       "      <th>RNA adapter</th>\n",
       "      <th>RIBO strandedness</th>\n",
       "      <th>RNA strandedness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE22001</td>\n",
       "      <td>SRP003554</td>\n",
       "      <td>SRX026871</td>\n",
       "      <td>SRX026872</td>\n",
       "      <td>28,29,30</td>\n",
       "      <td>12,12,12</td>\n",
       "      <td>TCGTATG</td>\n",
       "      <td>TCGTATG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE72064</td>\n",
       "      <td>SRP062407</td>\n",
       "      <td>SRX1149649</td>\n",
       "      <td>SRX1149668</td>\n",
       "      <td>28,29,30,31</td>\n",
       "      <td>12,12,12,12</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSE84112</td>\n",
       "      <td>SRP078005</td>\n",
       "      <td>SRX1900396</td>\n",
       "      <td>SRX1900402</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSE89011</td>\n",
       "      <td>SRP091889</td>\n",
       "      <td>SRX2255510</td>\n",
       "      <td>SRX2255511</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCT</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCT</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSE102890</td>\n",
       "      <td>SRP115915</td>\n",
       "      <td>SRX3110803</td>\n",
       "      <td>SRX3110807</td>\n",
       "      <td>29,30,31,32,33,34</td>\n",
       "      <td>12,12,12,13,13,13</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GSE99787</td>\n",
       "      <td>SRP108787</td>\n",
       "      <td>SRX2893684</td>\n",
       "      <td>SRX2893692</td>\n",
       "      <td>25,26,27,28</td>\n",
       "      <td>12,10,11,12</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         GEO        SRA RIBO sample  RNA sample       read lengths  \\\n",
       "0   GSE22001  SRP003554   SRX026871   SRX026872           28,29,30   \n",
       "1   GSE72064  SRP062407  SRX1149649  SRX1149668        28,29,30,31   \n",
       "2   GSE84112  SRP078005  SRX1900396  SRX1900402     26,27,28,29,30   \n",
       "3   GSE89011  SRP091889  SRX2255510  SRX2255511     26,27,28,29,30   \n",
       "4  GSE102890  SRP115915  SRX3110803  SRX3110807  29,30,31,32,33,34   \n",
       "5   GSE99787  SRP108787  SRX2893684  SRX2893692        25,26,27,28   \n",
       "\n",
       "             offsets           RIBO adapter            RNA adapter  \\\n",
       "0           12,12,12                TCGTATG                TCGTATG   \n",
       "1        12,12,12,12  TGGAATTCTCGGGTGCCAAGG  TGGAATTCTCGGGTGCCAAGG   \n",
       "2     12,12,12,12,12  TGGAATTCTCGGGTGCCAAGG  TGGAATTCTCGGGTGCCAAGG   \n",
       "3     12,12,12,12,12  AGATCGGAAGAGCACACGTCT  AGATCGGAAGAGCACACGTCT   \n",
       "4  12,12,12,13,13,13  TGGAATTCTCGGGTGCCAAGG  TGGAATTCTCGGGTGCCAAGG   \n",
       "5        12,10,11,12                   None                   None   \n",
       "\n",
       "  RIBO strandedness RNA strandedness  \n",
       "0           forward          forward  \n",
       "1           forward          forward  \n",
       "2           forward          forward  \n",
       "3           forward          forward  \n",
       "4           forward          forward  \n",
       "5           forward          reverse  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sheet_name = 'mm10'\n",
    "excel_df = pd.read_excel('../data/datasets_info.xlsx', sheet_name=sheet_name)\n",
    "excel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "config_out_dir = '/auto/cmb-panasas2/skchoudh/github_projects/ribocop-results/tools/spectre/configs'\n",
    "input_dir_root = '/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/'\n",
    "intron_bed = '/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.gffutils.intron.bed'\n",
    "chrom_sizes = '/home/cmb-panasas2/skchoudh/genomes/mm10/fasta/mm10.chrom.sizes'\n",
    "genome_fasta = '/home/cmb-panasas2/skchoudh/genomes/mm10/fasta/mm10.fa'\n",
    "gtf = '/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.annotation.gtf'\n",
    "star_index = '/home/cmb-panasas2/skchoudh/genomes/mm10/star_annotated/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_strandedness(strandedness):\n",
    "    strandedness_map = {'forward': 'fr-secondstrand',\n",
    "                         'reverse': 'fr-firststrand',\n",
    "                         'no': 'fr-unstranded'}\n",
    "    return strandedness_map[strandedness]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRP003554\n",
      "SRP062407\n",
      "SRP078005\n",
      "SRP091889\n",
      "SRP115915\n",
      "SRP108787\n"
     ]
    }
   ],
   "source": [
    "offset_file_map = {}\n",
    "for index, row in excel_df.iterrows():\n",
    "    read_lengths = list(row['read lengths'].strip().split(','))\n",
    "    offsets = row['offsets'].strip().split(',')\n",
    "    \n",
    "    zipped = zip(read_lengths, offsets)\n",
    "    targets = (',').join([(':').join(x) for x in zipped])\n",
    "    srp = row['SRA']\n",
    "    print(srp)\n",
    "    ribo_sample = row['RIBO sample']\n",
    "    rna_sample = row['RIBO sample']\n",
    "    input_dir = os.path.join(input_dir_root, srp+'_mouse')\n",
    "    rna_strand = convert_strandedness(row['RNA strandedness'])\n",
    "    ribo_strand = convert_strandedness(row['RIBO strandedness'])\n",
    "    config = dedent('''\n",
    "    INPUT_DIR = '{}'\n",
    "    GTF = '{}'\n",
    "    RIBO_STRAND_CUFFLINKS = '{}'\n",
    "    RNA_STRAND_CUFFLINKS = '{}'\n",
    "    OFFSETS = '\"{}\"'\n",
    "    ''').format(input_dir, gtf, ribo_strand, rna_strand, targets)\n",
    "    with open(os.path.join(config_out_dir, '{}_mouse.py'.format(srp)), 'w') as fh:\n",
    "        fh.write(config)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEO</th>\n",
       "      <th>SRA</th>\n",
       "      <th>RIBO sample</th>\n",
       "      <th>RNA sample</th>\n",
       "      <th>read lengths</th>\n",
       "      <th>offsets</th>\n",
       "      <th>RIBO adapter</th>\n",
       "      <th>RNA adapter</th>\n",
       "      <th>RIBO strandedness</th>\n",
       "      <th>RNA strandedness</th>\n",
       "      <th>AUC(RiboCop/RiboCode/RiboTaper/ORFscore)</th>\n",
       "      <th>metagene_coherence_ribocop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE35469</td>\n",
       "      <td>SRP010679</td>\n",
       "      <td>SRX118286</td>\n",
       "      <td>SRX118285</td>\n",
       "      <td>28,29,30</td>\n",
       "      <td>12,13,13</td>\n",
       "      <td>CTGTAGGCAC</td>\n",
       "      <td>CTGTAGGCAC</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>SRP029589</td>\n",
       "      <td>SRX345309</td>\n",
       "      <td>SRX345311</td>\n",
       "      <td>29,30,32</td>\n",
       "      <td>12,12,13</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSE73136</td>\n",
       "      <td>SRP063852</td>\n",
       "      <td>SRX1254413</td>\n",
       "      <td>SRX426378 (SRP028887, GSE49831)</td>\n",
       "      <td>28,29,30</td>\n",
       "      <td>12,12,12</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSE94454</td>\n",
       "      <td>SRP098789</td>\n",
       "      <td>SRX2536421</td>\n",
       "      <td>SRX2536426</td>\n",
       "      <td>28,30</td>\n",
       "      <td>12,13</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSE96716</td>\n",
       "      <td>SRP102021</td>\n",
       "      <td>SRX2647167</td>\n",
       "      <td>SRX2647164</td>\n",
       "      <td>28,29,30,31</td>\n",
       "      <td>12,12,12,12</td>\n",
       "      <td>TCGTATGCCGTCTTCTGCTTG</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GSE87328</td>\n",
       "      <td>SRP090415</td>\n",
       "      <td>SRX2189157</td>\n",
       "      <td>SRX2189154</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>AGATCGGAAGAGCACAC</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GSE89183</td>\n",
       "      <td>SRP092068</td>\n",
       "      <td>SRX2268404</td>\n",
       "      <td>SRX2268396</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCT</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GSE96714</td>\n",
       "      <td>SRP102020</td>\n",
       "      <td>SRX2647162</td>\n",
       "      <td>SRX2647160</td>\n",
       "      <td>29,30</td>\n",
       "      <td>12,12</td>\n",
       "      <td>TCGTATGCCGTCTTCTGCTTG</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GEO        SRA RIBO sample                       RNA sample  \\\n",
       "0  GSE35469  SRP010679   SRX118286                        SRX118285   \n",
       "1       NaN  SRP029589   SRX345309                        SRX345311   \n",
       "2  GSE73136  SRP063852  SRX1254413  SRX426378 (SRP028887, GSE49831)   \n",
       "3  GSE94454  SRP098789  SRX2536421                       SRX2536426   \n",
       "4  GSE96716  SRP102021  SRX2647167                       SRX2647164   \n",
       "5  GSE87328  SRP090415  SRX2189157                       SRX2189154   \n",
       "6  GSE89183  SRP092068  SRX2268404                       SRX2268396   \n",
       "7  GSE96714  SRP102020  SRX2647162                       SRX2647160   \n",
       "\n",
       "     read lengths         offsets           RIBO adapter        RNA adapter  \\\n",
       "0        28,29,30        12,13,13             CTGTAGGCAC         CTGTAGGCAC   \n",
       "1        29,30,32        12,12,13      CTGTAGGCACCATCAAT  CTGTAGGCACCATCAAT   \n",
       "2        28,29,30        12,12,12                   None               None   \n",
       "3           28,30           12,13      CTGTAGGCACCATCAAT  CTGTAGGCACCATCAAT   \n",
       "4     28,29,30,31     12,12,12,12  TCGTATGCCGTCTTCTGCTTG               None   \n",
       "5  26,27,28,29,30  12,12,12,12,12      CTGTAGGCACCATCAAT  AGATCGGAAGAGCACAC   \n",
       "6  26,27,28,29,30  12,12,12,12,12  AGATCGGAAGAGCACACGTCT               None   \n",
       "7           29,30           12,12  TCGTATGCCGTCTTCTGCTTG               None   \n",
       "\n",
       "  RIBO strandedness RNA strandedness  \\\n",
       "0           forward          forward   \n",
       "1           forward          forward   \n",
       "2           forward          forward   \n",
       "3           forward          forward   \n",
       "4           forward          forward   \n",
       "5           forward          reverse   \n",
       "6           forward          reverse   \n",
       "7           forward               no   \n",
       "\n",
       "   AUC(RiboCop/RiboCode/RiboTaper/ORFscore)  metagene_coherence_ribocop  \n",
       "0                                       NaN                         NaN  \n",
       "1                                       NaN                         NaN  \n",
       "2                                       NaN                         NaN  \n",
       "3                                       NaN                         NaN  \n",
       "4                                       NaN                         NaN  \n",
       "5                                       NaN                         NaN  \n",
       "6                                       NaN                         NaN  \n",
       "7                                       NaN                         NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_common_name_map = {'mm10': 'mouse', 'hg38': 'human'}\n",
    "\n",
    "sheet_name = 'hg38'\n",
    "excel_df = pd.read_excel('../data/datasets_info.xlsx', sheet_name=sheet_name)\n",
    "excel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "config_out_dir = '/auto/cmb-panasas2/skchoudh/github_projects/ribocop-results/tools/spectre/configs'\n",
    "input_dir_root = '/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/'\n",
    "gtf = '/home/cmb-panasas2/skchoudh/genomes/hg38/annotation/gencode.v25.annotation.gtf'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRP010679\n",
      "SRP029589\n",
      "SRP063852\n",
      "SRP098789\n",
      "SRP102021\n",
      "SRP090415\n",
      "SRP092068\n",
      "SRP102020\n"
     ]
    }
   ],
   "source": [
    "offset_file_map = {}\n",
    "for index, row in excel_df.iterrows():\n",
    "    read_lengths = list(row['read lengths'].strip().split(','))\n",
    "    offsets = row['offsets'].strip().split(',')\n",
    "    \n",
    "    zipped = zip(read_lengths, offsets)\n",
    "    targets = (',').join([(':').join(x) for x in zipped])\n",
    "    srp = row['SRA']\n",
    "    print(srp)\n",
    "    ribo_sample = row['RIBO sample']\n",
    "    rna_sample = row['RIBO sample']\n",
    "    input_dir = os.path.join(input_dir_root, srp+'_human')\n",
    "    rna_strand = convert_strandedness(row['RNA strandedness'])\n",
    "    ribo_strand = convert_strandedness(row['RIBO strandedness'])\n",
    "    config = dedent('''\n",
    "    INPUT_DIR = '{}'\n",
    "    GTF = '{}'\n",
    "    RIBO_STRAND_CUFFLINKS = '{}'\n",
    "    RNA_STRAND_CUFFLINKS = '{}'\n",
    "    OFFSETS = '\"{}\"'\n",
    "    ''').format(input_dir, gtf, ribo_strand, rna_strand, targets)\n",
    "    with open(os.path.join(config_out_dir, '{}_human.py'.format(srp)), 'w') as fh:\n",
    "        fh.write(config)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3 (pysradb)",
   "language": "python",
   "name": "pysradb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
