{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from textwrap import dedent\n",
    "import os\n",
    "import errno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "1944416 2199837 3133323 4000644 5137597 6866109 7824961 8587911 9909020 0560884"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mkdir_p(path):\n",
    "    \"\"\"Python version mkdir -p\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    path : string\n",
    "           Path to directory to create\n",
    "    \"\"\"\n",
    "    if path:\n",
    "        try:\n",
    "            os.makedirs(path)\n",
    "        except OSError as exc:  # Python >2.5\n",
    "            if exc.errno == errno.EEXIST and os.path.isdir(path):\n",
    "                pass\n",
    "            else:\n",
    "                raise\n",
    "def symlink_force(target, link_name):\n",
    "    try:\n",
    "        os.symlink(target, link_name)\n",
    "    except OSError as e:\n",
    "        if e.errno == errno.EEXIST:\n",
    "            os.remove(link_name)\n",
    "            os.symlink(target, link_name)\n",
    "        else:\n",
    "            raise e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEO</th>\n",
       "      <th>SRA</th>\n",
       "      <th>RIBO sample</th>\n",
       "      <th>RNA sample</th>\n",
       "      <th>read lengths</th>\n",
       "      <th>offsets</th>\n",
       "      <th>RIBO adapter</th>\n",
       "      <th>RNA adapter</th>\n",
       "      <th>RIBO strandedness</th>\n",
       "      <th>RNA strandedness</th>\n",
       "      <th>AUC(RiboCop/RiboCode/RiboTaper/ORFscore)</th>\n",
       "      <th>metagene_coherence_ribocop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE35469</td>\n",
       "      <td>SRP010679</td>\n",
       "      <td>SRX118286</td>\n",
       "      <td>SRX118285</td>\n",
       "      <td>28,29,30</td>\n",
       "      <td>12,13,13</td>\n",
       "      <td>CTGTAGGCAC</td>\n",
       "      <td>CTGTAGGCAC</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>SRP029589</td>\n",
       "      <td>SRX345309</td>\n",
       "      <td>SRX345311</td>\n",
       "      <td>29,30,32</td>\n",
       "      <td>12,12,13</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSE73136</td>\n",
       "      <td>SRP063852</td>\n",
       "      <td>SRX1254413</td>\n",
       "      <td>SRX426378 (SRP028887, GSE49831)</td>\n",
       "      <td>28,29,30</td>\n",
       "      <td>12,12,12</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSE94454</td>\n",
       "      <td>SRP098789</td>\n",
       "      <td>SRX2536421</td>\n",
       "      <td>SRX2536426</td>\n",
       "      <td>28,30</td>\n",
       "      <td>12,13</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSE96716</td>\n",
       "      <td>SRP102021</td>\n",
       "      <td>SRX2647167</td>\n",
       "      <td>SRX2647164</td>\n",
       "      <td>28,29,30,31</td>\n",
       "      <td>12,12,12,12</td>\n",
       "      <td>TCGTATGCCGTCTTCTGCTTG</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GSE87328</td>\n",
       "      <td>SRP090415</td>\n",
       "      <td>SRX2189157</td>\n",
       "      <td>SRX2189154</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>AGATCGGAAGAGCACAC</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GSE89183</td>\n",
       "      <td>SRP092068</td>\n",
       "      <td>SRX2268404</td>\n",
       "      <td>SRX2268396</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCT</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GSE96714</td>\n",
       "      <td>SRP102020</td>\n",
       "      <td>SRX2647162</td>\n",
       "      <td>SRX2647160</td>\n",
       "      <td>29,30</td>\n",
       "      <td>12,12</td>\n",
       "      <td>TCGTATGCCGTCTTCTGCTTG</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GEO        SRA RIBO sample                       RNA sample  \\\n",
       "0  GSE35469  SRP010679   SRX118286                        SRX118285   \n",
       "1       NaN  SRP029589   SRX345309                        SRX345311   \n",
       "2  GSE73136  SRP063852  SRX1254413  SRX426378 (SRP028887, GSE49831)   \n",
       "3  GSE94454  SRP098789  SRX2536421                       SRX2536426   \n",
       "4  GSE96716  SRP102021  SRX2647167                       SRX2647164   \n",
       "5  GSE87328  SRP090415  SRX2189157                       SRX2189154   \n",
       "6  GSE89183  SRP092068  SRX2268404                       SRX2268396   \n",
       "7  GSE96714  SRP102020  SRX2647162                       SRX2647160   \n",
       "\n",
       "     read lengths         offsets           RIBO adapter        RNA adapter  \\\n",
       "0        28,29,30        12,13,13             CTGTAGGCAC         CTGTAGGCAC   \n",
       "1        29,30,32        12,12,13      CTGTAGGCACCATCAAT  CTGTAGGCACCATCAAT   \n",
       "2        28,29,30        12,12,12                   None               None   \n",
       "3           28,30           12,13      CTGTAGGCACCATCAAT  CTGTAGGCACCATCAAT   \n",
       "4     28,29,30,31     12,12,12,12  TCGTATGCCGTCTTCTGCTTG               None   \n",
       "5  26,27,28,29,30  12,12,12,12,12      CTGTAGGCACCATCAAT  AGATCGGAAGAGCACAC   \n",
       "6  26,27,28,29,30  12,12,12,12,12  AGATCGGAAGAGCACACGTCT               None   \n",
       "7           29,30           12,12  TCGTATGCCGTCTTCTGCTTG               None   \n",
       "\n",
       "  RIBO strandedness RNA strandedness  \\\n",
       "0           forward          forward   \n",
       "1           forward          forward   \n",
       "2           forward          forward   \n",
       "3           forward          forward   \n",
       "4           forward          forward   \n",
       "5           forward          reverse   \n",
       "6           forward          reverse   \n",
       "7           forward               no   \n",
       "\n",
       "   AUC(RiboCop/RiboCode/RiboTaper/ORFscore)  metagene_coherence_ribocop  \n",
       "0                                       NaN                         NaN  \n",
       "1                                       NaN                         NaN  \n",
       "2                                       NaN                         NaN  \n",
       "3                                       NaN                         NaN  \n",
       "4                                       NaN                         NaN  \n",
       "5                                       NaN                         NaN  \n",
       "6                                       NaN                         NaN  \n",
       "7                                       NaN                         NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_common_name_map = {'mm10': 'mouse', 'hg38': 'human'}\n",
    "\n",
    "sheet_name = 'hg38'\n",
    "excel_df = pd.read_excel('../data/datasets_info.xlsx', sheet_name=sheet_name)\n",
    "excel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir_root = '/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/'\n",
    "genome_name = 'hg38'\n",
    "genome_base_path = '/home/cmb-panasas2/skchoudh/genomes/hg38/rpbp_index_mrna'\n",
    "gtf = '/home/cmb-panasas2/skchoudh/genomes/hg38/annotation/gencode.v25.annotation.gtf'\n",
    "genome_fasta = '/home/cmb-panasas2/skchoudh/genomes/hg38/fasta/hg38.fa'\n",
    "ribosomal_index = '/home/cmb-panasas2/skchoudh/genomes/hg38/rpbp_index_rrna'\n",
    "star_index = '/home/cmb-panasas2/skchoudh/genomes/hg38/rpbp_star_index'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP010679_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP010679_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP029589_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP029589_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP063852_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP063852_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP098789_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP098789_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP102021_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP102021_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP090415_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP090415_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP092068_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP092068_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP102020_human/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/hg38/SRP102020_human/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n"
     ]
    }
   ],
   "source": [
    "for index, row in excel_df.iterrows():\n",
    "    lengths = list(row['read lengths'].strip().split(','))\n",
    "    offsets = row['offsets'].strip().split(',')\n",
    "    srp = row['SRA']\n",
    "    ribo_sample = row['RIBO sample']\n",
    "    rna_sample = row['RIBO sample']\n",
    "    input_dir = os.path.join(input_dir_root, srp+'_human')\n",
    "    riboseq_data = os.path.join(input_dir, 'rpbp_results')\n",
    "\n",
    "    symlink_dir = os.path.join(riboseq_data, 'without-rrna-mapping')\n",
    "    mkdir_p(symlink_dir)\n",
    "    ribo_file = os.path.join(input_dir, 'RIBO_CCDS_lengths.bam')\n",
    "    rna_file = os.path.join(input_dir, 'RNA_CCDS_lengths.bam')\n",
    "    \n",
    "    ribo_file_symlink = os.path.join(symlink_dir, 'RIBO_CCDS_lengths.alignments-only.bam')\n",
    "    rna_file_symlink = os.path.join(symlink_dir, 'RNA_CCDS_lengths.alignments-only.bam')\n",
    "    \n",
    "    symlink_force(rna_file, rna_file_symlink)\n",
    "    print(rna_file, rna_file_symlink)\n",
    "    symlink_force(ribo_file, ribo_file_symlink)\n",
    "\n",
    "    hg38_config = \"\"\"\n",
    "    genome_name: {}\n",
    "    genome_base_path: {}\n",
    "    gtf: {}\n",
    "    fasta: {}\n",
    "    ribosomal_index: {}\n",
    "    star_index: {}\n",
    "\n",
    "    #adapter_file: /path/to/rp-bp/riboseq-adapters.fa\n",
    "\n",
    "    riboseq_data: {}\n",
    "\n",
    "    riboseq_samples:\n",
    "        RIBO_CCDS_lengths: \n",
    "        RNA_CCDS_lengths:\n",
    "\n",
    "\n",
    "    riboseq_biological_replicates:\n",
    "        # these must still be the same names from riboseq_samples\n",
    "        RIBO_CCDS_lengths:\n",
    "            - RIBO_CCDS_lengths\n",
    "        RNA_CCDS_lengths:\n",
    "            - RNA_CCDS_lengths\n",
    "\n",
    "    note: alignments-only\n",
    "\n",
    "    orf_note: atg-only\n",
    "\n",
    "    #models_base: /path/to/rp-bp/models\n",
    "    max_uncalled: 1\n",
    "    pre_trim_left: 0\n",
    "    adapter_sequence: null\n",
    "    quality_format: sanger\n",
    "\n",
    "\n",
    "    sjdb_overhang: 50\n",
    "    align_intron_minc: 20\n",
    "    align_intron_maxc: 100000\n",
    "    out_filter_intron_motifsc: RemoveNoncanonicalUnannotated\n",
    "    out_filter_mismatch_n_maxc: 1\n",
    "    out_filter_mismatch_n_over_l_maxc: 0.04\n",
    "    out_filter_typec: BySJout\n",
    "    out_sam_attributesc: [AS, NH, HI, nM, MD]\n",
    "\n",
    "    # METAGENE PERIODICITY OPTIONS, OPTIONAL\n",
    "\n",
    "    # the number of bases upstream of the translation initiation site to begin \n",
    "    # constructing the metagene profile. default: 50\n",
    "    metagene_profile_start_upstream: 50 \n",
    "\n",
    "    # the number of bases downstream of the translation initiation site to end \n",
    "    # the metagene profile. default: 20\n",
    "    metagene_profile_start_downstream: 20\n",
    "\n",
    "    # the number of bases upstream of the translation termination site to begin \n",
    "    # constructing the metagene profile. default: 50\n",
    "    metagene_profile_end_upstream: 50\n",
    "\n",
    "    # the number of bases downstream of the translation termination site to end \n",
    "    # the metagene profile. default: 20\n",
    "    metagene_profile_end_downstream: 20\n",
    "\n",
    "    # the position, relative to the translation initiation site, to begin \n",
    "    # calculating periodicity Bayes factors (inclusive)\n",
    "    periodic_offset_start: -20\n",
    "\n",
    "    # the position, relative to the translation initiation site, to stop \n",
    "    # calculating periodicity Bayes factors (inclusive)\n",
    "    periodic_offset_end: 0\n",
    "\n",
    "    # the length of the profile to use in the models. \n",
    "\n",
    "    # N.B. metagene_profile_length + periodic_offset_end must be consistent with \n",
    "    # the length of the extracted metagene profile\n",
    "    metagene_profile_length: 21\n",
    "\n",
    "    # the number of iterations to use for each chain in the MCMC sampling. The \n",
    "    # first half of the iterations are discarded as burn-in samples. All of \n",
    "    # the remaining samples are used to estimate the posterior distributions. That \n",
    "    # is, we do not use thinning. default: 500\n",
    "\n",
    "    # N.B. Increasing the number of iterations will increase the running time of\n",
    "    # the estimation roughly linearly.\n",
    "    metagene_profile_iterations: 500\n",
    "\n",
    "    # SHARED MARKOV CHAIN MONTE CARLO OPTIONS, OPTIONAL\n",
    "\n",
    "    # These affect the MCMC both for estimating metagene profile periodicity and \n",
    "    # ORF translation Bayes factors.\n",
    "\n",
    "    # the random seed for the MCMC sampling. default: 8675309\n",
    "    seed: 8675309 \n",
    "\n",
    "    # the number of chains to use in the MCMC sampling. default: 2\n",
    "\n",
    "    # N.B. Increases in the number of chains should improve the quality of\n",
    "    # estimation (especially reducing variance), but will increase the running\n",
    "    # time proportionally. For example, changing the number of chains from 2 to\n",
    "    # 4 will roughly double the running time of estimation.\n",
    "    chains: 2\n",
    "\n",
    "    # PERIODICITY AND OFFSET OPTIONS, OPTIONAL\n",
    "\n",
    "    # FIXED LENGTHS AND OFFSETS\n",
    "    # By default, the Rp-Bp pipelines use the periodicity estimation technique to\n",
    "    # select periodic read lengths and P-site offsets. However, fixed lengths and\n",
    "    # offsets can be specified with the following options.\n",
    "\n",
    "    # if this variable is present in the config file with any value, then the \n",
    "    # estimated periodic read lengths and offsets will not be used. Instead, \n",
    "    # fixed values given by lengths and offsets (below) will be used.\n",
    "\n",
    "    # uncomment this line to use the fixed lengths and offsets\n",
    "    use_fixed_lengths: yes\n",
    "\n",
    "    # a list of read lengths which will be used for creating the profiles if the \n",
    "    # use_fixed_lengths option is given. Presumably, these are lengths that have \n",
    "    # periodic metagene profiles.\n",
    "\n",
    "    # This value is ignored if 'use_fixed_lengths' is not set.\n",
    "    lengths: {}\n",
    "\n",
    "    # the P-site offset to use for each read length specifed by --lengths if the \n",
    "    # use_fixed_lengths option is given. The number of offsets must match the \n",
    "    # number of lengths, and they are assumed to match. For example lengths of \n",
    "    # [26, 29] with offsets [9, 12] means only reads of lengths 26 bp and 29 bp \n",
    "    # will be used to create the profiles. The 26 bp reads will be shifted by \n",
    "    # 9 bp in the 5' direction, while reads of length 29 bp will be shifted by 12 bp.\n",
    "\n",
    "    # This value is ignored if 'use_fixed_lengths' is not set.\n",
    "    offsets: {}\n",
    "\n",
    "    # if fixed lengths are not used, the minimum number of reads for a particular \n",
    "    # length in the filtered genome profile. Read lengths with fewer than this \n",
    "    # number of reads will not be used. default: 1000\n",
    "    #min_metagene_profile_count: 1000\n",
    "\n",
    "    # N.B. This value is set artificially low for the example to work!\n",
    "    min_metagene_profile_count: 10\n",
    "\n",
    "    # if fixed lengths are not used...\n",
    "    # if max_metagene_profile_bayes_factor_var is not None, then this is taken as a \n",
    "    # hard threshold on the estimated Bayes factor mean. \n",
    "\n",
    "    # If min_metagene_bf_likelihood is given, then this is taken as \n",
    "    # the boundary value; that is, a profile is \"periodic\" if:\n",
    "    #\n",
    "    #       [P(bf > min_metagene_bf_mean)] > min_metagene_bf_likelihood\n",
    "    #\n",
    "    # If both max_metagene_bf_var and min_metagene_bf_likelihood are None, then this \n",
    "    # is taken as a hard threshold on the mean for selecting periodic read lengths.\n",
    "    #\n",
    "    # If both max_metagene_bf_var and min_metagene_bf_likelihood are given, then both \n",
    "    # filters will be applied and the result will be the intersection.\n",
    "    min_metagene_bf_mean: 5\n",
    "\n",
    "    # if given, then this is taken as a hard threshold on the estimated Bayes factor \n",
    "    # variance. default: None, i.e., this filter is not used. (null in yaml) \n",
    "    max_metagene_bf_var: null\n",
    "\n",
    "    # if given, then this is taken a threshold on the likelihood of periodicity \n",
    "    # (see min_metagene_bf_mean description for more details). default: 0.5\n",
    "    min_metagene_bf_likelihood: 0.5\n",
    "\n",
    "    # SMOOTHING OPTIONS, OPTIONAL\n",
    "\n",
    "    # the fraction of the profile to use for smoothing within LOWESS. default: 0.2\n",
    "    smoothing_fraction: 0.2\n",
    "\n",
    "    # the number of reweighting iterations to use within LOWESS. Please see the\n",
    "    # statsmodels documentation for a detailed description of this parameter.\n",
    "    # default: 0\n",
    "    smoothing_reweighting_iterations: 0\n",
    "\n",
    "    # if this value is greater than 0, then ORFs whose length (in nucleotides) \n",
    "    # is less than this value will not be smoothed, and neither the Bayes factor \n",
    "    # estimates nor the chi-square p-value will be calcualted. default: 0\n",
    "    min_orf_length: 0\n",
    "\n",
    "    # if this value is greater than 0, then ORFs whose length (in nucleotides) \n",
    "    # is greater than this value will not be smoothed, and neither the Bayes factor \n",
    "    # estimates nor the chi-square p-value will be calcualted. default: 0\n",
    "    max_orf_length: 0\n",
    "\n",
    "    # ORFs for which the number of in-frame reads is less than this value will not\n",
    "    # be smoothed, and neither the Bayes factor estimates nor the chi-square\n",
    "    # p-value will be calculated. default: 5\n",
    "    min_signal: 5\n",
    "\n",
    "\n",
    "    # BAYES FACTOR ESTIMATION OPTIONS, OPTIONAL\n",
    "\n",
    "    # if this flag is in the config file with any value, then only the Rp-chi \n",
    "    # pipeline will be performed; namely, the translation models will not be fit \n",
    "    # to the data, and the posterior distributions will not be estimated.\n",
    "\n",
    "    # uncomment this line to only run the Rp-chi pipeline\n",
    "    #chi_square_only: yes\n",
    "\n",
    "    # the number of iterations to use for each chain in the MCMC sampling. The first \n",
    "    # half of the iterations are discarded as burn-in samples. All of the remaining \n",
    "    # samples are used to estimate the posterior distributions. (That is, we do not \n",
    "    # use thinning.) default: 200\n",
    "\n",
    "    # N.B. Increasing the number of iterations will increase the running time of\n",
    "    # the estimation roughly linearly.\n",
    "    translation_iterations: 200\n",
    "\n",
    "    # SELECTION PREDICTED ORFS OPTIONS, OPTIONAL\n",
    "\n",
    "    # the minimum value for the estimated Bayes factor mean to \"predict\" that an \n",
    "    # ORF is translated. This value is used in conjunction with both min_bf_mean\n",
    "    # and min_bf_likelihood.\n",
    "    #\n",
    "    # If max_bf_var is a positive value, then this is taken as a hard threshold on \n",
    "    # the estimated Bayes factor mean. ORFs must meet both the min_bf_mean and\n",
    "    # max_bf_var filters to be selected as \"translated.\"\n",
    "    #\n",
    "    # If min_bf_likelihood is given, then this is taken as the boundary value; that\n",
    "    # is, an ORF is selected as \"translated\" if:\n",
    "    #\n",
    "    #   [P(bf > min_bf_mean)] > min_bf_likelihood\n",
    "    #\n",
    "    # If both max_bf_var and min_bf_likelihood are None, then this is taken as a\n",
    "    # hard threshold on the mean for selecting translated ORFs.\n",
    "    #\n",
    "    # If both max_bf_var and min_bf_likelihood are given, then both filters will \n",
    "    # be applied and the result will be the intersection.\n",
    "    #\n",
    "    # default: 5\n",
    "    min_bf_mean: 5\n",
    "\n",
    "    # the maximum value value for the estimated Bayes factor variance to \"predict\" \n",
    "    # that an ORF is translated. ORFs must meet both the min_bf_mean and max_bf_var \n",
    "    # filters to be predicted. See the description for min_bf_mean for more \n",
    "    # details. default: null (i.e., this filter is not used by default)\n",
    "    max_bf_var: null\n",
    "\n",
    "    # the minimum probability of the BF exceeding min_bf_mean to select an ORF as\n",
    "    # translated. See the description for min_bf_mean for more details. default: 0.5\n",
    "    min_bf_likelihood: 0.5\n",
    "\n",
    "    # for the chi-square test, this value is first Bonferroni corrected \n",
    "    # based on the number of ORFs which pass the smoothing filters. It is \n",
    "    # then used as the significance threshold to select translated ORFs. default: 0.01\n",
    "    chisq_significance_level: 0.01\n",
    "    \"\"\".format(genome_name, genome_base_path, gtf,\n",
    "               genome_fasta, ribosomal_index, star_index,\n",
    "               riboseq_data, lengths, offsets)\n",
    "    hg38_config = hg38_config.replace(\"'\",\"\")\n",
    "    with open('/home/cmb-panasas2/skchoudh/github_projects/ribocop-results/tools/rpbp/config_{}_human.yaml'.format(srp), 'w') as fh:\n",
    "        fh.write(dedent(hg38_config))\n",
    "    to_write = 'run-all-rpbp-instances config_{}_human.yaml --overwrite --num-cpus 16 --logging-level INFO --log-file log.alignments-only_{}.txt --keep-intermediate-files'.format(srp, srp)\n",
    "    with open('/home/cmb-panasas2/skchoudh/github_projects/ribocop-results/tools/rpbp/run_rpbp_{}_human.sh'.format(srp), 'w') as fh:\n",
    "        fh.write(to_write)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ribo_file_symlink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ribo_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEO</th>\n",
       "      <th>SRA</th>\n",
       "      <th>RIBO sample</th>\n",
       "      <th>RNA sample</th>\n",
       "      <th>read lengths</th>\n",
       "      <th>offsets</th>\n",
       "      <th>RIBO adapter</th>\n",
       "      <th>RNA adapter</th>\n",
       "      <th>RIBO strandedness</th>\n",
       "      <th>RNA strandedness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE22001</td>\n",
       "      <td>SRP003554</td>\n",
       "      <td>SRX026871</td>\n",
       "      <td>SRX026872</td>\n",
       "      <td>28,29,30</td>\n",
       "      <td>12,12,12</td>\n",
       "      <td>TCGTATG</td>\n",
       "      <td>TCGTATG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE72064</td>\n",
       "      <td>SRP062407</td>\n",
       "      <td>SRX1149649</td>\n",
       "      <td>SRX1149668</td>\n",
       "      <td>28,29,30,31</td>\n",
       "      <td>12,12,12,12</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSE84112</td>\n",
       "      <td>SRP078005</td>\n",
       "      <td>SRX1900396</td>\n",
       "      <td>SRX1900402</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSE89011</td>\n",
       "      <td>SRP091889</td>\n",
       "      <td>SRX2255510</td>\n",
       "      <td>SRX2255511</td>\n",
       "      <td>26,27,28,29,30</td>\n",
       "      <td>12,12,12,12,12</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCT</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCT</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSE102890</td>\n",
       "      <td>SRP115915</td>\n",
       "      <td>SRX3110803</td>\n",
       "      <td>SRX3110807</td>\n",
       "      <td>29,30,31,32,33,34</td>\n",
       "      <td>12,12,12,13,13,13</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>TGGAATTCTCGGGTGCCAAGG</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GSE99787</td>\n",
       "      <td>SRP108787</td>\n",
       "      <td>SRX2893684</td>\n",
       "      <td>SRX2893692</td>\n",
       "      <td>25,26,27,28</td>\n",
       "      <td>12,10,11,12</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         GEO        SRA RIBO sample  RNA sample       read lengths  \\\n",
       "0   GSE22001  SRP003554   SRX026871   SRX026872           28,29,30   \n",
       "1   GSE72064  SRP062407  SRX1149649  SRX1149668        28,29,30,31   \n",
       "2   GSE84112  SRP078005  SRX1900396  SRX1900402     26,27,28,29,30   \n",
       "3   GSE89011  SRP091889  SRX2255510  SRX2255511     26,27,28,29,30   \n",
       "4  GSE102890  SRP115915  SRX3110803  SRX3110807  29,30,31,32,33,34   \n",
       "5   GSE99787  SRP108787  SRX2893684  SRX2893692        25,26,27,28   \n",
       "\n",
       "             offsets           RIBO adapter            RNA adapter  \\\n",
       "0           12,12,12                TCGTATG                TCGTATG   \n",
       "1        12,12,12,12  TGGAATTCTCGGGTGCCAAGG  TGGAATTCTCGGGTGCCAAGG   \n",
       "2     12,12,12,12,12  TGGAATTCTCGGGTGCCAAGG  TGGAATTCTCGGGTGCCAAGG   \n",
       "3     12,12,12,12,12  AGATCGGAAGAGCACACGTCT  AGATCGGAAGAGCACACGTCT   \n",
       "4  12,12,12,13,13,13  TGGAATTCTCGGGTGCCAAGG  TGGAATTCTCGGGTGCCAAGG   \n",
       "5        12,10,11,12                   None                   None   \n",
       "\n",
       "  RIBO strandedness RNA strandedness  \n",
       "0           forward          forward  \n",
       "1           forward          forward  \n",
       "2           forward          forward  \n",
       "3           forward          forward  \n",
       "4           forward          forward  \n",
       "5           forward          reverse  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_common_name_map = {'mm10': 'mouse', 'hg38': 'human'}\n",
    "\n",
    "sheet_name = 'mm10'\n",
    "excel_df = pd.read_excel('../data/datasets_info.xlsx', sheet_name=sheet_name)\n",
    "excel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir_root = '/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/'\n",
    "genome_name = 'mm10'\n",
    "genome_base_path = '/home/cmb-panasas2/skchoudh/genomes/mm10/rpbp_index_mrna'\n",
    "gtf = '/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.annotation.gtf'\n",
    "genome_fasta = '/home/cmb-panasas2/skchoudh/genomes/mm10/fasta/mm10.fa'\n",
    "ribosomal_index = '/home/cmb-panasas2/skchoudh/genomes/mm10/rpbp_index_rrna'\n",
    "star_index = '/home/cmb-panasas2/skchoudh/genomes/mm10/rpbp_star_index'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP003554_mouse/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP003554_mouse/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP062407_mouse/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP062407_mouse/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP078005_mouse/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP078005_mouse/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP091889_mouse/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP091889_mouse/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP115915_mouse/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP115915_mouse/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n",
      "/home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP108787_mouse/RNA_CCDS_lengths.bam /home/cmb-panasas2/skchoudh/benchmark/gene_level/mm10/SRP108787_mouse/rpbp_results/without-rrna-mapping/RNA_CCDS_lengths.alignments-only.bam\n"
     ]
    }
   ],
   "source": [
    "for index, row in excel_df.iterrows():\n",
    "    lengths = list(row['read lengths'].strip().split(','))\n",
    "    offsets = row['offsets'].strip().split(',')\n",
    "    srp = row['SRA']\n",
    "    ribo_sample = row['RIBO sample']\n",
    "    rna_sample = row['RIBO sample']\n",
    "    input_dir = os.path.join(input_dir_root, srp+'_mouse')\n",
    "    riboseq_data = os.path.join(input_dir, 'rpbp_results')\n",
    "\n",
    "    symlink_dir = os.path.join(riboseq_data, 'without-rrna-mapping')\n",
    "    mkdir_p(symlink_dir)\n",
    "    ribo_file = os.path.join(input_dir, 'RIBO_CCDS_lengths.bam')\n",
    "    rna_file = os.path.join(input_dir, 'RNA_CCDS_lengths.bam')\n",
    "    \n",
    "    ribo_file_symlink = os.path.join(symlink_dir, 'RIBO_CCDS_lengths.alignments-only.bam')\n",
    "    rna_file_symlink = os.path.join(symlink_dir, 'RNA_CCDS_lengths.alignments-only.bam')\n",
    "    \n",
    "    symlink_force(rna_file, rna_file_symlink)\n",
    "    print(rna_file, rna_file_symlink)\n",
    "    symlink_force(ribo_file, ribo_file_symlink)\n",
    "    \n",
    "    riboseq_data = os.path.join(input_dir, 'rpbp_results')\n",
    "    config = \"\"\"\n",
    "    genome_name: {}\n",
    "    genome_base_path: {}\n",
    "    gtf: {}\n",
    "    fasta: {}\n",
    "    ribosomal_index: {}\n",
    "    star_index: {}\n",
    "\n",
    "    #adapter_file: /path/to/rp-bp/riboseq-adapters.fa\n",
    "\n",
    "    riboseq_data: {}\n",
    "\n",
    "    riboseq_samples:\n",
    "        RIBO_CCDS_lengths: \n",
    "        RNA_CCDS_lengths:\n",
    "\n",
    "\n",
    "    riboseq_biological_replicates:\n",
    "        # these must still be the same names from riboseq_samples\n",
    "        RIBO_CCDS_lengths:\n",
    "            - RIBO_CCDS_lengths\n",
    "        RNA_CCDS_lengths:\n",
    "            - RNA_CCDS_lengths\n",
    "\n",
    "    note: alignments-only\n",
    "\n",
    "    orf_note: atg-only\n",
    "\n",
    "    #models_base: /path/to/rp-bp/models\n",
    "    max_uncalled: 1\n",
    "    pre_trim_left: 0\n",
    "    adapter_sequence: null\n",
    "    quality_format: sanger\n",
    "\n",
    "\n",
    "    sjdb_overhang: 50\n",
    "    align_intron_minc: 20\n",
    "    align_intron_maxc: 100000\n",
    "    out_filter_intron_motifsc: RemoveNoncanonicalUnannotated\n",
    "    out_filter_mismatch_n_maxc: 1\n",
    "    out_filter_mismatch_n_over_l_maxc: 0.04\n",
    "    out_filter_typec: BySJout\n",
    "    out_sam_attributesc: [AS, NH, HI, nM, MD]\n",
    "\n",
    "    # METAGENE PERIODICITY OPTIONS, OPTIONAL\n",
    "\n",
    "    # the number of bases upstream of the translation initiation site to begin \n",
    "    # constructing the metagene profile. default: 50\n",
    "    metagene_profile_start_upstream: 50 \n",
    "\n",
    "    # the number of bases downstream of the translation initiation site to end \n",
    "    # the metagene profile. default: 20\n",
    "    metagene_profile_start_downstream: 20\n",
    "\n",
    "    # the number of bases upstream of the translation termination site to begin \n",
    "    # constructing the metagene profile. default: 50\n",
    "    metagene_profile_end_upstream: 50\n",
    "\n",
    "    # the number of bases downstream of the translation termination site to end \n",
    "    # the metagene profile. default: 20\n",
    "    metagene_profile_end_downstream: 20\n",
    "\n",
    "    # the position, relative to the translation initiation site, to begin \n",
    "    # calculating periodicity Bayes factors (inclusive)\n",
    "    periodic_offset_start: -20\n",
    "\n",
    "    # the position, relative to the translation initiation site, to stop \n",
    "    # calculating periodicity Bayes factors (inclusive)\n",
    "    periodic_offset_end: 0\n",
    "\n",
    "    # the length of the profile to use in the models. \n",
    "\n",
    "    # N.B. metagene_profile_length + periodic_offset_end must be consistent with \n",
    "    # the length of the extracted metagene profile\n",
    "    metagene_profile_length: 21\n",
    "\n",
    "    # the number of iterations to use for each chain in the MCMC sampling. The \n",
    "    # first half of the iterations are discarded as burn-in samples. All of \n",
    "    # the remaining samples are used to estimate the posterior distributions. That \n",
    "    # is, we do not use thinning. default: 500\n",
    "\n",
    "    # N.B. Increasing the number of iterations will increase the running time of\n",
    "    # the estimation roughly linearly.\n",
    "    metagene_profile_iterations: 500\n",
    "\n",
    "    # SHARED MARKOV CHAIN MONTE CARLO OPTIONS, OPTIONAL\n",
    "\n",
    "    # These affect the MCMC both for estimating metagene profile periodicity and \n",
    "    # ORF translation Bayes factors.\n",
    "\n",
    "    # the random seed for the MCMC sampling. default: 8675309\n",
    "    seed: 8675309 \n",
    "\n",
    "    # the number of chains to use in the MCMC sampling. default: 2\n",
    "\n",
    "    # N.B. Increases in the number of chains should improve the quality of\n",
    "    # estimation (especially reducing variance), but will increase the running\n",
    "    # time proportionally. For example, changing the number of chains from 2 to\n",
    "    # 4 will roughly double the running time of estimation.\n",
    "    chains: 2\n",
    "\n",
    "    # PERIODICITY AND OFFSET OPTIONS, OPTIONAL\n",
    "\n",
    "    # FIXED LENGTHS AND OFFSETS\n",
    "    # By default, the Rp-Bp pipelines use the periodicity estimation technique to\n",
    "    # select periodic read lengths and P-site offsets. However, fixed lengths and\n",
    "    # offsets can be specified with the following options.\n",
    "\n",
    "    # if this variable is present in the config file with any value, then the \n",
    "    # estimated periodic read lengths and offsets will not be used. Instead, \n",
    "    # fixed values given by lengths and offsets (below) will be used.\n",
    "\n",
    "    # uncomment this line to use the fixed lengths and offsets\n",
    "    use_fixed_lengths: yes\n",
    "\n",
    "    # a list of read lengths which will be used for creating the profiles if the \n",
    "    # use_fixed_lengths option is given. Presumably, these are lengths that have \n",
    "    # periodic metagene profiles.\n",
    "\n",
    "    # This value is ignored if 'use_fixed_lengths' is not set.\n",
    "    lengths: {}\n",
    "\n",
    "    # the P-site offset to use for each read length specifed by --lengths if the \n",
    "    # use_fixed_lengths option is given. The number of offsets must match the \n",
    "    # number of lengths, and they are assumed to match. For example lengths of \n",
    "    # [26, 29] with offsets [9, 12] means only reads of lengths 26 bp and 29 bp \n",
    "    # will be used to create the profiles. The 26 bp reads will be shifted by \n",
    "    # 9 bp in the 5' direction, while reads of length 29 bp will be shifted by 12 bp.\n",
    "\n",
    "    # This value is ignored if 'use_fixed_lengths' is not set.\n",
    "    offsets: {}\n",
    "\n",
    "    # if fixed lengths are not used, the minimum number of reads for a particular \n",
    "    # length in the filtered genome profile. Read lengths with fewer than this \n",
    "    # number of reads will not be used. default: 1000\n",
    "    #min_metagene_profile_count: 1000\n",
    "\n",
    "    # N.B. This value is set artificially low for the example to work!\n",
    "    min_metagene_profile_count: 10\n",
    "\n",
    "    # if fixed lengths are not used...\n",
    "    # if max_metagene_profile_bayes_factor_var is not None, then this is taken as a \n",
    "    # hard threshold on the estimated Bayes factor mean. \n",
    "\n",
    "    # If min_metagene_bf_likelihood is given, then this is taken as \n",
    "    # the boundary value; that is, a profile is \"periodic\" if:\n",
    "    #\n",
    "    #       [P(bf > min_metagene_bf_mean)] > min_metagene_bf_likelihood\n",
    "    #\n",
    "    # If both max_metagene_bf_var and min_metagene_bf_likelihood are None, then this \n",
    "    # is taken as a hard threshold on the mean for selecting periodic read lengths.\n",
    "    #\n",
    "    # If both max_metagene_bf_var and min_metagene_bf_likelihood are given, then both \n",
    "    # filters will be applied and the result will be the intersection.\n",
    "    min_metagene_bf_mean: 5\n",
    "\n",
    "    # if given, then this is taken as a hard threshold on the estimated Bayes factor \n",
    "    # variance. default: None, i.e., this filter is not used. (null in yaml) \n",
    "    max_metagene_bf_var: null\n",
    "\n",
    "    # if given, then this is taken a threshold on the likelihood of periodicity \n",
    "    # (see min_metagene_bf_mean description for more details). default: 0.5\n",
    "    min_metagene_bf_likelihood: 0.5\n",
    "\n",
    "    # SMOOTHING OPTIONS, OPTIONAL\n",
    "\n",
    "    # the fraction of the profile to use for smoothing within LOWESS. default: 0.2\n",
    "    smoothing_fraction: 0.2\n",
    "\n",
    "    # the number of reweighting iterations to use within LOWESS. Please see the\n",
    "    # statsmodels documentation for a detailed description of this parameter.\n",
    "    # default: 0\n",
    "    smoothing_reweighting_iterations: 0\n",
    "\n",
    "    # if this value is greater than 0, then ORFs whose length (in nucleotides) \n",
    "    # is less than this value will not be smoothed, and neither the Bayes factor \n",
    "    # estimates nor the chi-square p-value will be calcualted. default: 0\n",
    "    min_orf_length: 0\n",
    "\n",
    "    # if this value is greater than 0, then ORFs whose length (in nucleotides) \n",
    "    # is greater than this value will not be smoothed, and neither the Bayes factor \n",
    "    # estimates nor the chi-square p-value will be calcualted. default: 0\n",
    "    max_orf_length: 0\n",
    "\n",
    "    # ORFs for which the number of in-frame reads is less than this value will not\n",
    "    # be smoothed, and neither the Bayes factor estimates nor the chi-square\n",
    "    # p-value will be calculated. default: 5\n",
    "    min_signal: 5\n",
    "\n",
    "\n",
    "    # BAYES FACTOR ESTIMATION OPTIONS, OPTIONAL\n",
    "\n",
    "    # if this flag is in the config file with any value, then only the Rp-chi \n",
    "    # pipeline will be performed; namely, the translation models will not be fit \n",
    "    # to the data, and the posterior distributions will not be estimated.\n",
    "\n",
    "    # uncomment this line to only run the Rp-chi pipeline\n",
    "    #chi_square_only: yes\n",
    "\n",
    "    # the number of iterations to use for each chain in the MCMC sampling. The first \n",
    "    # half of the iterations are discarded as burn-in samples. All of the remaining \n",
    "    # samples are used to estimate the posterior distributions. (That is, we do not \n",
    "    # use thinning.) default: 200\n",
    "\n",
    "    # N.B. Increasing the number of iterations will increase the running time of\n",
    "    # the estimation roughly linearly.\n",
    "    translation_iterations: 200\n",
    "\n",
    "    # SELECTION PREDICTED ORFS OPTIONS, OPTIONAL\n",
    "\n",
    "    # the minimum value for the estimated Bayes factor mean to \"predict\" that an \n",
    "    # ORF is translated. This value is used in conjunction with both min_bf_mean\n",
    "    # and min_bf_likelihood.\n",
    "    #\n",
    "    # If max_bf_var is a positive value, then this is taken as a hard threshold on \n",
    "    # the estimated Bayes factor mean. ORFs must meet both the min_bf_mean and\n",
    "    # max_bf_var filters to be selected as \"translated.\"\n",
    "    #\n",
    "    # If min_bf_likelihood is given, then this is taken as the boundary value; that\n",
    "    # is, an ORF is selected as \"translated\" if:\n",
    "    #\n",
    "    #   [P(bf > min_bf_mean)] > min_bf_likelihood\n",
    "    #\n",
    "    # If both max_bf_var and min_bf_likelihood are None, then this is taken as a\n",
    "    # hard threshold on the mean for selecting translated ORFs.\n",
    "    #\n",
    "    # If both max_bf_var and min_bf_likelihood are given, then both filters will \n",
    "    # be applied and the result will be the intersection.\n",
    "    #\n",
    "    # default: 5\n",
    "    min_bf_mean: 5\n",
    "\n",
    "    # the maximum value value for the estimated Bayes factor variance to \"predict\" \n",
    "    # that an ORF is translated. ORFs must meet both the min_bf_mean and max_bf_var \n",
    "    # filters to be predicted. See the description for min_bf_mean for more \n",
    "    # details. default: null (i.e., this filter is not used by default)\n",
    "    max_bf_var: null\n",
    "\n",
    "    # the minimum probability of the BF exceeding min_bf_mean to select an ORF as\n",
    "    # translated. See the description for min_bf_mean for more details. default: 0.5\n",
    "    min_bf_likelihood: 0.5\n",
    "\n",
    "    # for the chi-square test, this value is first Bonferroni corrected \n",
    "    # based on the number of ORFs which pass the smoothing filters. It is \n",
    "    # then used as the significance threshold to select translated ORFs. default: 0.01\n",
    "    chisq_significance_level: 0.01\n",
    "    \"\"\".format(genome_name, genome_base_path, gtf,\n",
    "               genome_fasta, ribosomal_index, star_index,\n",
    "               riboseq_data, lengths, offsets)\n",
    "    config = config.replace(\"'\",\"\")\n",
    "    with open('/home/cmb-panasas2/skchoudh/github_projects/ribocop-results/tools/rpbp/config_{}_mouse.yaml'.format(srp), 'w') as fh:\n",
    "        fh.write(dedent(config))\n",
    "    to_write = 'run-all-rpbp-instances config_{}_mouse.yaml --overwrite --num-cpus 16 --logging-level INFO --log-file log.alignments-only_{}_mouse.txt --keep-intermediate-files'.format(srp, srp)\n",
    "    with open('/home/cmb-panasas2/skchoudh/github_projects/ribocop-results/tools/rpbp/run_rpbp_{}_mouse.sh'.format(srp), 'w') as fh:\n",
    "        fh.write(to_write)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:riboraptor]",
   "language": "python",
   "name": "conda-env-riboraptor-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
